### PR TITLE
fix(lambda-python): bundle asset files correctly

### LINKED ISSUE #18301: ‼️ (aws-lambda-python): SAM - CDK integration is broken in CDK V2.4.0
### What is the problem?

There is a new change in CDK Version 2.4.0 that makes the bundled assets to be created under the folder structure as following:
```
├── <asset root>
│   └── asset-input
│       ├── asset files
```
 
The issue is SAM expects to get the asset location in metadata property `aws:asset:path`, so SAM can read the Lambda Function source code, so the customers can do local testing, but in the new Version, the value of the added metadata is the path till the Asset root directory (till the <Asset Hash>), although it should be (<Asset Hash>/asset-input)

For an application that contains a PythonFunction, This is the directory structure after running `cdk synth` using version 2.4.0
```
147dda696921:cdk_app melasmar$ cdk --version
2.4.0 (build 993f14d)
147dda696921:cdk_app melasmar$ cd cdk.out
147dda696921:cdk.out melasmar$ tree
.
├── CDKV2SupportDemoStack.assets.json
├── CDKV2SupportDemoStack.template.json
├── asset.993168cfa75b295eeea55bf603340284b9be46ebc079e4965f2c16f5470efda5
│   └── asset-input
│       ├── __init__.py
│       └── app.py
├── cdk.out
├── manifest.json
└── tree.json
```
and this is the synthesized template:
```
"MyFunction3BAA72D1": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
              ............   
        },
      "DependsOn": [
        "MyFunctionServiceRole3C357FF2"
      ],
      "Metadata": {
        "aws:cdk:path": "CDKV2SupportDemoStack/MyFunction/Resource",
        "aws:asset:path": "asset.993168cfa75b295eeea55bf603340284b9be46ebc079e4965f2c16f5470efda5",
        "aws:asset:is-bundled": true,
        "aws:asset:property": "Code"
      }
```

and this is the output in case of using CDK version 2.3.0
```
147dda696921:cdk_app melasmar$ cdk --version
2.3.0 (build beaa5b2)
147dda696921:cdk_app melasmar$ cd cdk.out
147dda696921:cdk.out melasmar$ tree
.
├── CDKV2SupportDemoStack.assets.json
├── CDKV2SupportDemoStack.template.json
├── asset.8f47b59bc8d67146c1b3fdd46c077e64cf2e0021298beb44f424619eeaaad8d0
│   ├── __init__.py
│   └── app.py
├── cdk.out
├── manifest.json
└── tree.json
```
and this is the synthesized template:
```
"MyFunction3BAA72D1": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
         .........
      },
      "DependsOn": [
        "MyFunctionServiceRole3C357FF2"
      ],
      "Metadata": {
        "aws:cdk:path": "CDKV2SupportDemoStack/MyFunction/Resource",
        "aws:asset:path": "asset.8f47b59bc8d67146c1b3fdd46c077e64cf2e0021298beb44f424619eeaaad8d0",
        "aws:asset:is-bundled": true,
        "aws:asset:property": "Code"
      }
    },
```

### Reproduction Steps

- create a simple CDK stack that contains a PythonFunction.
- run `cdk synth`
- check the synthesized template, and the bundled assets directory

### What did you expect to happen?

The value of the added `aws:asset:path` metadata property should be the location of the source code of the bundled assets.

### What actually happened?

The value of the added `aws:asset:path` metadata property is the root asset directory.

### CDK CLI Version

2.4.0

### Framework Version

_No response_

### Node.js Version

v14.15.4

### OS

mac os

### Language

Typescript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by setu4993**
Asset files are incorrectly being bundled under the `asset-input` directory instead of the root of the bundle.

To also copy over hidden files (https://github.com/aws/aws-cdk/pull/18306#discussion_r780186564), I switched from using `-R` to `-a` based on what I found on [SO](https://stackoverflow.com/a/13020113) and the [man page](https://linux.die.net/man/1/cp). (`-a` is equivalent to `-dR`.)

Fixes #18301 and @chrispykim's comment: https://github.com/aws/aws-cdk/pull/18082#issuecomment-1008442363.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by setu4993** [comment]
cc @njlynch

**Comment by setu4993** [comment]
I tested this locally in a Docker container and it bundles correctly for me. I'm not very familiar with how to check the contents of the bundling directory, so not sure how to add a test for it. I think it makes sense to get a release out that fixes this ASAP and then add a test later.

**Review by otaviomacedo — DISMISSED** [review]

File: packages/@aws-cdk/aws-lambda-python/lib/bundling.ts [hunk 1]
```diff
@@ -87,7 +87,7 @@ export class Bundling implements CdkBundlingOptions {
     if (packaging.dependenciesFile) {
       bundlingCommands.push(`python -m pip install -r ${DependenciesFile.PIP} -t ${options.outputDir}`);
     }
-    bundlingCommands.push(`cp -R ${options.inputDir}/ ${options.outputDir}`);
+    bundlingCommands.push(`cp -a ${options.inputDir}/* ${options.outputDir}`);
```
Comment (njlynch) [inline packages/@aws-cdk/aws-lambda-python/lib/bundling.ts:90]: Including the `*` still drops hidden files. This needs to be removed.

```suggestion
    bundlingCommands.push(`cp -a ${options.inputDir}/ ${options.outputDir}`);
```

**Review by njlynch — CHANGES_REQUESTED** [review]
Do we have a reproducible example for what's still breaking for people? I worry we're throwing darts at the map (or flags on the CLI) and hoping it will fix it without truly understanding the root cause.

**Review by corymhall — APPROVED** [review]

File: packages/@aws-cdk/aws-lambda-python/lib/bundling.ts (hunk 1 above)
Comment (setu4993) [reply packages/@aws-cdk/aws-lambda-python/lib/bundling.ts]: Urm, I added a validation for it within the test, which passes.

**Comment by setu4993** [comment]
> Do we have a reproducible example for what's still breaking for people? I worry we're throwing darts at the map (or flags on the CLI) and hoping it will fix it without truly understanding the root cause.

That's what I added assertions to validate. I'm confident those would fail for what's currently on the release versions because it would include an extra layer in between.

**Pull request merged**